src/Interpreters/ConcurrentHashJoin.h

```diff
@@ -18,18 +18,26 @@ namespace DB
 struct SelectQueryInfo;
 
 /**
- * Can run addBlockToJoin() parallelly to speedup the join process. On test, it almose linear speedup by
- * the degree of parallelism.
+ * The default `HashJoin` is not thread-safe for inserting the right table’s rows; thus, it is done on a single thread.
+ * When the right table is large, the join process is too slow.
  *
- * The default HashJoin is not thread safe for inserting right table's rows and run it in a single thread. When
- * the right table is large, the join process is too slow.
- *
- * We create multiple HashJoin instances here. In addBlockToJoin(), one input block is split into multiple blocks
- * corresponding to the HashJoin instances by hashing every row on the join keys. And make a guarantee that every HashJoin
- * instance is written by only one thread.
- *
- * When come to the left table matching, the blocks from left table are alse split into different HashJoin instances.
+ * `ConcurrentHashJoin` can run `addBlockToJoin()` concurrently to speed up the join process. On the test, it scales almost linearly.
+ * For that, we create multiple `HashJoin` instances. In `addBlockToJoin()`, one input block is split into multiple blocks
+ * corresponding to the `HashJoin` instances by hashing every row on the join keys. In particular, each `HashJoin` instance has its own hash map
+ * that stores a unique set of keys. Also, `addBlockToJoin()` calls are done under mutex to guarantee
+ * that every `HashJoin` instance is written only from one thread at a time.
  *
+ * When matching the left table, the input blocks are also split by hash and routed to corresponding `HashJoin` instances.
+ * This introduces some noticeable overhead compared to the `hash` join algorithm that doesn’t have to split. Then,
+ * we introduced the following optimization. On the probe stage, we want to have the same execution as for the `hash` join algorithm,
+ * i.e., we want to have a single shared hash map that we will read from each thread. No splitting of blocks is required.
+ * We should somehow divide this shared hash map between threads so that we can still execute the build stage concurrently.
+ * The idea is to use a two-level hash map and distribute its buckets between threads. Namely, we will calculate the same hash
+ * that the hash map calculates, map it to the bucket number, and then take this number modulo the number of threads. This way,
+ * upon build phase completion, we will have thread #0 having a hash map with only buckets {#0, #threads_num, #threads_num*2, ...},
+ * thread #1 with only buckets {#1, #threads_num+1, #threads_num*2+1, ...} and so on. To form the resulting hash map,
```

Comment:
Why do we still need the ConcurrentHashJoin class with this approach? If the map in HashJoin is already two-level, can’t we dispatch rows directly by the maps at the second level? The current approach seems more like a three-level map, doesn’t it? Is this because removing the extra level would require significantly more refactoring, or is this design necessary for some specific reason?